core/embed/rust/src/ui/model_tt/component/homescreen/render.rs

```diff
@@ -78,6 +82,99 @@ const RED_IDX: usize = 0;
 const GREEN_IDX: usize = 1;
 const BLUE_IDX: usize = 2;
 
+pub trait HomescreenDecompressor {
+    fn get_height(&self) -> i16;
+    fn decompress(&mut self);
+    fn get_data(&mut self) -> &mut BufferJpeg;
+}
+
+pub struct HomescreenJpeg<'i> {
+    pub output: BufferOutput<'i>,
+    pub jdec: Option<JDEC<'i, 'i>>,
+}
+
+impl<'i> HomescreenJpeg<'i> {
+    pub fn new(input: &'i mut BufferInput<'i>) -> Self {
+        Self {
+            output: BufferOutput::new(unsafe { get_jpeg_buffer(0, true) }, WIDTH, 16),
+            jdec: JDEC::new(input, unsafe {
+                get_jpeg_work_buffer(0, true).buffer.as_mut_slice()
+            })
+            .ok(),
+        }
+    }
+}
+
+impl<'i> HomescreenDecompressor for HomescreenJpeg<'i> {
+    fn get_height(&self) -> i16 {
+        if let Some(dec) = self.jdec.as_ref() {
+            return dec.mcu_height();
+        }
+        1
+    }
+
+    fn decompress(&mut self) {
+        self.jdec.as_mut().map(|dec| dec.decomp(&mut self.output));
+    }
+
+    fn get_data(&mut self) -> &mut BufferJpeg {
+        self.output.buffer()
+    }
+}
+
+pub struct HomescreenToif<'i> {
+    pub output: BufferOutput<'i>,
+    pub decomp_context: UzlibContext<'i>,
+    line: i16,
+}
+
+impl<'i> HomescreenToif<'i> {
+    pub fn new(context: UzlibContext<'i>) -> Self {
+        Self {
+            output: BufferOutput::new(unsafe { get_jpeg_buffer(0, true) }, WIDTH, 16),
+            decomp_context: context,
+            line: 0,
+        }
+    }
+}
+
+impl<'i> HomescreenDecompressor for HomescreenToif<'i> {
+    fn get_height(&self) -> i16 {
+        1
+    }
+
+    fn decompress(&mut self) {
+        //self.jdec.as_mut().map(|dec| dec.decomp(&mut self.output));
+        // SAFETY: Aligning to u8 slice is safe, because the original slice is aligned
+        // to 16 bits, therefore there are also no residuals (prefix/suffix).
+        // The data in the slices are integers, so these are valid for both u16
+        // and u8.
+        if self.line >= HOMESCREEN_TOIF_Y_OFFSET
+            && self.line < HOMESCREEN_TOIF_Y_OFFSET + HOMESCREEN_TOIF_SIZE
+        {
+            let (_, workbuf, _) = unsafe { self.output.buffer().buffer.align_to_mut::<u8>() };
+            unwrap!(self.decomp_context.uncompress(
```

Comment:
Will this RSOD on malformed/short image data? Might be better to just draw black pixels in this case, that way device still remains usable?

Reply:
https://github.com/trezor/trezor-firmware/commit/efad3596e8d2fc3c208e13384990bda439334289